Framework/Muon/src/LoadAndApplyMuonDetectorGrouping.cpp

```diff
@@ -0,0 +1,366 @@
+
+  std::string analysisGrp("Analysis Options");
+  setPropertyGroup("RebinArgs", analysisGrp);
+  setPropertyGroup("TimeOffset", analysisGrp);
+  setPropertyGroup("SummedPeriods", analysisGrp);
+  setPropertyGroup("SubtractedPeriods", analysisGrp);
+  setPropertyGroup("DeadTimeTable", analysisGrp);
+}
+
+/**
+ * Performs validation of inputs to the algorithm.
+ * - Check the input workspace and WorkspaceGroup are not the same
+ */
+std::map<std::string, std::string>
+LoadAndApplyMuonDetectorGrouping::validateInputs() {
+  std::map<std::string, std::string> errors;
+
+  if (!this->isDefault("WorkspaceGroup")) {
+
+    WorkspaceGroup_sptr groupedWS = getProperty("WorkspaceGroup");
+    Workspace_sptr inputWS = getProperty("InputWorkspace");
+
+    if (groupedWS && groupedWS->getName() == inputWS->getName()) {
+      errors["WorkspaceGroup"] = "The WorkspaceGroup should not have "
+                                 "the same name as InputWorkspace.";
+    }
+  }
+
+  return errors;
+}
+
+void LoadAndApplyMuonDetectorGrouping::exec() {
+  this->setRethrows(true);
+
+  AnalysisOptions options = setDefaultOptions();
+  options.grouping = loadGroupsAndPairs();
+
+  Workspace_sptr inputWS = getProperty("InputWorkspace");
+
+  checkDetectorIDsInWorkspace(options.grouping, inputWS);
+
+  WorkspaceGroup_sptr groupedWS = getProperty("WorkspaceGroup");
+  if (!groupedWS) {
+    groupedWS = addGroupedWSWithDefaultName(inputWS);
+  }
+
+  // Perform the analysis specified in options
+  addGroupingToADS(options, inputWS, groupedWS);
+  addPairingToADS(options, inputWS, groupedWS);
+
+  if (getProperty("ApplyAsymmetryToGroups")) {
+    options.plotType = PlotType::Asymmetry;
+    addGroupingToADS(options, inputWS, groupedWS);
+  }
+  addGroupingInformationToADS(options.grouping);
+}
+
+// Checks that the detector IDs in grouping are in the workspace
+void LoadAndApplyMuonDetectorGrouping::checkDetectorIDsInWorkspace(
+    API::Grouping &grouping, Workspace_sptr workspace) {
+  bool check =
+      MuonAlgorithmHelper::checkGroupDetectorsInWorkspace(grouping, workspace);
+  if (!check) {
+    g_log.error("One or more detector IDs specified in the groups is not "
+                "contained in the InputWorkspace");
+    throw std::runtime_error(
+        "One or more detector IDs specified in the groups is not "
+        "contained in the InputWorkspace");
+  }
+}
+
+/**
+ * Adds an empty WorkspaceGroup to the ADS with a name
+ * that it would have if created by the MuonAnalysis GUI.
+ */
+WorkspaceGroup_sptr
+LoadAndApplyMuonDetectorGrouping::addGroupedWSWithDefaultName(
+    Workspace_sptr workspace) {
+  auto &ads = AnalysisDataService::Instance();
+  std::string groupedWSName = MuonAlgorithmHelper::getRunLabel(workspace);
+
+  WorkspaceGroup_sptr groupedWS;
+  if (ads.doesExist(groupedWSName)) {
+    if ((groupedWS = boost::dynamic_pointer_cast<WorkspaceGroup>(
+             ads.retrieve(groupedWSName)))) {
+      return groupedWS;
+    }
+  }
+
```

Comment:
then you can place the next 2 lines in an else. That way everything uses the same return and its a bit neater